File: frontend/src/api/consistency.ts
import apiClient from './client';

export interface ConsistencyCheck {
  id: string;
  source_a_id: string;
  source_b_id: string;
  field_mapping_id?: string;
  status: string;
  results?: Record<string, unknown>;
  summary?: string;
  started_at?: string;
  completed_at?: string;
  created_at: string;
  updated_at: string;
}

export interface Anomaly {
  id: string;
  check_id: string;
  source_a_id: string;
  source_b_id: string;
  field_name: string;
  severity: string;
  status: string;
  source_a_value?: unknown;
  source_b_value?: unknown;
  expected_value?: unknown;
  actual_value?: unknown;
  root_cause?: string;
  resolution?: string;
  created_at: string;
  updated_at: string;
}

export interface RunCheckRequest {
  source_a_id: string;
  source_b_id: string;
  field_mapping_id?: string;
}

export interface ResolveAnomalyRequest {
  resolution: string;
  status: 'resolved' | 'dismissed';
}

export const consistencyApi = {
  runCheck: async (data: RunCheckRequest): Promise<ConsistencyCheck> => {
    const res = await apiClient.post('/consistency/checks', data);
    return res.data;
  },

  listChecks: async (skip = 0, limit = 50): Promise<{ items: ConsistencyCheck[]; total: number }> => {
    const res = await apiClient.get('/consistency/checks', { params: { skip, limit } });
    return res.data;
  },

  getCheck: async (id: string): Promise<ConsistencyCheck> => {
    const res = await apiClient.get(`/consistency/checks/${id}`);
    return res.data;
  },

  listAnomalies: async (
    skip = 0,
    limit = 50,
    severity?: string,
    status?: string
  ): Promise<{ items: Anomaly[]; total: number }> => {
    const res = await apiClient.get('/consistency/anomalies', {
      params: { skip, limit, severity, status },
    });
    return res.data;
  },

  resolveAnomaly: async (id: string, data: ResolveAnomalyRequest): Promise<Anomaly> => {
    const res = await apiClient.put(`/consistency/anomalies/${id}/resolve`, data);
    return res.data;
  },
};

File: backend/app/schemas/consistency.py
"""Pydantic schemas for consistency checks and anomalies."""

from pydantic import BaseModel, Field
from typing import Optional, Any
from datetime import datetime
import uuid


class RunCheckRequest(BaseModel):
    source_a_id: str = Field(..., description="UUID of the first data source")
    source_b_id: str = Field(..., description="UUID of the second data source")
    field_mapping_id: Optional[str] = None


class ConsistencyCheckResponse(BaseModel):
    id: uuid.UUID
    source_a_id: uuid.UUID
    source_b_id: uuid.UUID
    field_mapping_id: Optional[uuid.UUID] = None
    status: str
    results: Optional[dict[str, Any]] = None
    summary: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ConsistencyCheckListResponse(BaseModel):
    items: list[ConsistencyCheckResponse]
    total: int


class AnomalyResponse(BaseModel):
    id: uuid.UUID
    check_id: uuid.UUID
    source_a_id: uuid.UUID
    source_b_id: uuid.UUID
    field_name: str
    severity: str
    status: str
    source_a_value: Optional[Any] = None
    source_b_value: Optional[Any] = None
    expected_value: Optional[Any] = None
    actual_value: Optional[Any] = None
    root_cause: Optional[str] = None
    resolution: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AnomalyListResponse(BaseModel):
    items: list[AnomalyResponse]
    total: int


class ResolveAnomalyRequest(BaseModel):
    resolution: str = Field(..., min_length=1)
    status: str = Field(default="resolved", pattern="^(resolved|dismissed)$")